### DISCUSSION TITLE
Ways to set a "good" time-based termination policy based on problem complexity?

### DISCUSSION THREAD

**Discussion by greyhairredbear** [TimefoldAI/timefold-solver#1027] [General]
As already hinted [in another discussion](https://github.com/TimefoldAI/timefold-solver/discussions/917#discussioncomment-9873218), I have a question regarding termination configurations. 

Our users might try to solve problems of different complexity - from a few vehicles/loads to hundreds of loads/dozens of vehicles. For more complex problems, users are also prepared to wait for a longer time (e.g. an hour) for planning to finish, while this is not the case for simpler problems (only a couple of loads on one vehicle). 

Is there a way to set a timeout based on the complexity of the problem instance? I know that the benchmarker has some value for problem size - would it be the way to go to determine "good" timeouts for various problem sizes and then interpolate or use a step function for different complexity levels? Also, is the problem size part of the public API of Timefold?

**Comment by triceo** [comment]
The short answer is that there is no way to set a timeout based on the solution itself. Not directly.
The unimproved time termination is closest to what you need - at least you can benchmark and guesstimate how much time it would take for a particular dataset to flatline.

Arguably if you already know how long they're willing to wait for, just set that as the maximum time spent, and set unimproved time spent to a fraction of that.

**Reply by greyhairredbear** [reply]
> The unimproved time termination is closest to what you need - at least you can benchmark and guesstimate how much time it would take for a particular dataset to flatline.

Ok, so basically what I've guessed. Just thought that the `ProblemSizeStatistics` record could have been helpful for that. But I guess calculating the search space size manually beforehand and then adjusting the timeout depending on that should be possible in my case (though it will require a different data type than `Int` 😅 - still might be able to scale the timeout based on that). 

> Arguably if you already know how long they're willing to wait for, just set that as the maximum time spent, and set unimproved time spent to a fraction of that.

As mentioned, this is where my main problem lies. Depending on the problem, this value might just be a few seconds (e.g. just plan a single vehicle with 10 loads to 10 different locations) or even an hour (hundreds of loads to different locations).
